### PR TITLE
Update iotedged to return image pull failed message

### PR DESCRIPTION
**Pull request by philipktlin**
Update iotedged to return more details in response when image pull failed.
Cases tested:
1. invalid image name
2. invalid image host
3. network issue
4. unauthenticated request

### CODE REVIEW

**Comment by myagley** [comment]
I think we might be able to just report the message from the docker error, instead of having the beginning part? Something like: 

```
Output from iotedeged API endpoint for Create Module with invalid image name:

*   Trying 172.17.0.1...
* TCP_NODELAY set
* Connected to 172.17.0.1 (172.17.0.1) port 8080 (#0)
> POST /modules/?api-version=2018-06-28 HTTP/1.1
> Host: 172.17.0.1:8080
> User-Agent: curl/7.58.0
> Content-Type:application/json
> Accept:application/json
> Content-Length: 169
>
* upload completely sent off: 169 out of 169 bytes
< HTTP/1.1 404 Not Found
< content-type: application/json
< content-length: 117
< date: Thu, 11 Oct 2018 18:14:41 GMT
<
* Connection #0 to host 172.17.0.1 left intact
{"message":"manifest for mcr.microsoft.com/xazureiotedge-simulated-temperature-sensor:1.0 not found"}

Output from iotedeged API endpoint for Create Module with invalid image host:

*   Trying 172.17.0.1...
* TCP_NODELAY set
* Connected to 172.17.0.1 (172.17.0.1) port 8080 (#0)
> POST /modules/?api-version=2018-06-28 HTTP/1.1
> Host: 172.17.0.1:8080
> User-Agent: curl/7.58.0
> Content-Type:application/json
> Accept:application/json
> Content-Length: 169
>
* upload completely sent off: 169 out of 169 bytes
< HTTP/1.1 500 Internal Server Error
< content-type: application/json
< content-length: 150
< date: Thu, 11 Oct 2018 18:18:08 GMT
<
* Connection #0 to host 172.17.0.1 left intact
{"message":"Get https://xmcr.microsoft.com/v2/: dial tcp: lookup xmcr.microsoft.com on 127.0.0.53:53: no such host"}

Output from iotedged API endpoint for Create Module with missing authentication information:

*   Trying 172.17.0.1...
* TCP_NODELAY set
* Connected to 172.17.0.1 (172.17.0.1) port 8080 (#0)
> POST /modules/?api-version=2018-06-28 HTTP/1.1
> Host: 172.17.0.1:8080
> User-Agent: curl/7.58.0
> Content-Type:application/json
> Accept:application/json
> Content-Length: 169
>
* upload completely sent off: 169 out of 169 bytes
< HTTP/1.1 500 Internal Server Error
< content-type: application/json
< content-length: 176
< date: Thu, 11 Oct 2018 18:21:31 GMT
<
* Connection #0 to host 172.17.0.1 left intact
{"message":"Get https://phlinedgecontainerregistry.azurecr.io/v2/csharpfunction/manifests/0.0.1-amd64: unauthorized: authentication required"}

Output from iotedged API endpoint for Create Module with network issue (use iptables to block access to container registry):

*   Trying 172.17.0.1...
* TCP_NODELAY set
* Connected to 172.17.0.1 (172.17.0.1) port 8080 (#0)
> POST /modules/?api-version=2018-06-28 HTTP/1.1
> Host: 172.17.0.1:8080
> User-Agent: curl/7.58.0
> Content-Type:application/json
> Accept:application/json
> Content-Length: 169
>
* upload completely sent off: 169 out of 169 bytes
< HTTP/1.1 500 Internal Server Error
< content-type: application/json
< content-length: 207
< date: Thu, 11 Oct 2018 20:40:42 GMT
<
* Connection #0 to host 172.17.0.1 left intact
{"message":"Get https://phlinedgecontainerregistry.azurecr.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"}
```

File: edgelet/edgelet-docker/src/error.rs [hunk 1]
```diff
@@ -127,16 +146,20 @@ impl From<DockerError<serde_json::Value>> for Error {
...
+            DockerError::ApiError(error) => match error.code {
+                StatusCode::NOT_FOUND => match Error::get_message(error) {
+                    Ok(message) => Error::from(ErrorKind::NotFound(message)),
+                    Err(error) => Error::from(ErrorKind::DockerRuntime(DockerError::ApiError(error)))
+                },
+                StatusCode::CONFLICT =>
+                    Error::from(ErrorKind::Conflict),
+                StatusCode::NOT_MODIFIED =>
+                    Error::from(ErrorKind::NotModified),
+                _ => match Error::get_message(error) {
```
Comment (myagley) [inline edgelet/edgelet-docker/src/error.rs]:
```rust
_ => Error::get_message(error)
  .map(|message| Error::from(ErrorKind::FormattedDockerRuntime(message)))
  .unwrap_or_else(|e| Error::from(ErrorKind::DockerRuntime(DockerError::ApiError(e)))
```

may be more idiomatic

File: edgelet/edgelet-docker/src/error.rs [hunk 2]
```diff
@@ -127,16 +149,21 @@ impl From<DockerError<serde_json::Value>> for Error {
...
-            DockerError::ApiError(ref error) if error.code == StatusCode::NOT_MODIFIED => {
-                Error::from(ErrorKind::NotModified)
-            }
-            _ => Error::from(ErrorKind::DockerRuntime(err)),
+            DockerError::ApiError(error) => match error.code {
+                StatusCode::NOT_FOUND => match get_message(error) {
+                    Ok(message) => Error::from(ErrorKind::NotFound(message)),
+                    Err(error) => {
+                        Error::from(ErrorKind::DockerRuntime(DockerError::ApiError(error)))
+                    }
```
Comment (myagley) [inline edgelet/edgelet-docker/src/error.rs]: Can we do the same `map`/`unwrap_or_else` here (like the clause below?

Comment (philipktlin) [reply edgelet/edgelet-docker/src/error.rs]: thanks.

**Pull request merged**